### DISCUSSION TITLE
Before-Command-Exec hook? (terminal shell integration)

### DISCUSSION THREAD

**Discussion by christianparpart** [PowerShell/PowerShell#15933] [Q&A]
Hi,

I am used to be able to hook into the shell in order to better integrate the shell with the running terminal.

So how I can hook into the prompt I already found out, i.e. saving the script block of the current prompt, creating a new prompt function, doing my job and also call the saved script block).

But now I am also looking for a way to hook into the moment after the user has hit <kbd>Enter</kbd> to accept that input but before the input (command) is being executed.

That's possible in other shells (zsh, bash, fish, ...) but I could not find a solution for PowerShell just yet, is there?

Many thanks,
Christian Parpart.

p.s.: Why I am interested in this is because I need to be able to send some custom VT sequences to the host as some kind of notification system that a new command output has started.

**Comment by vexx32** [comment]
You can register events to `$ExecutionContext.InvokeCommand.PreCommandLookupAction` ([example](https://twitter.com/dissectmalware/status/1077818433016922112?lang=en)) to add logic to the command resolution itself, and you can register events to `$ExecutionContext.InvokeCommand.PostCommandLookupAction` to do things after the command to run has been resolved.

Barebones example with the latter:

```ps1
$ExecutionContext.InvokeCommand.PostCommandLookupAction = {
    param($Command, $CommandName, $CommandOrigin)
    Write-Host "Running $CommandName"
}
```

**Reply by christianparpart** [reply]
Oh yes. Many thanks. Will be trying that out ASAP.
